### PR TITLE
Optimizing getting parents by given type

### PR DESCRIPTION
**Pull request by sivajik34**
We can optimize type's  parents info  by storing into an array,so that next time onwards we can return this array with out calculating whole logic.

### CODE REVIEW

**Comment by adragus-inviqa** [comment]
You shouldn't use underscores for private/protected members.

Do you have some numbers? How much time does this save?

**Comment by sivajik34** [comment]
@adragus-inviqa 
Magento core team using underscores.
i don't know exactly how much time it will save or not.maybe useful to store into array without calculating again whole thing.

**Comment by adragus-inviqa** [comment]
Yeah, they shouldn't be using underscores. Must be some legacy code. But I'll wait for them to confirm.

Whenever you optimise stuff, you need to come up with numbers, to actually _prove_ that your change is needed. Is it saving a fraction of a second or 2 minutes? 

Every refactoring can introduce bugs, so your PR needs to be justified. With numbers.

**Comment by sivajik34** [comment]
@adragus-inviqa  i agree. can you give some idea to check the improvement.
I mean some tools,script to get actual numbers?
i think i can use profiler.

File: lib/internal/Magento/Framework/Code/Reader/ClassReader.php [hunk 1]
```diff
@@ -8,6 +8,11 @@
 class ClassReader implements ClassReaderInterface
 {
     /**
+     * @var array
+     */
+    protected $_classNameParents;
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/Code/Reader/ClassReader.php]: Please avoid using underscores in property names, it is a legacy standard.
There is no need for this property to be protected as we do not encourage inheritance-based APIs

File: lib/internal/Magento/Framework/Code/Reader/ClassReader.php [hunk 2]
```diff
@@ -56,27 +61,32 @@ public function getConstructor($className)
...
-                if ($parentInterfaces) {
-                    $result = array_values(array_diff($interfaces, $parentInterfaces));
-                } else {
-                    $result = array_values($interfaces);
-                }
-            }
-            array_unshift($result, $parentClass);
+        if (isset($this->_classNameParents[$className])) {
+            return $this->_classNameParents[$className];
         } else {
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/Code/Reader/ClassReader.php]: Looks like there is no more need for the `else`

File: lib/internal/Magento/Framework/ObjectManager/Config/Compiled.php [hunk 3]
```diff
@@ -120,6 +120,7 @@ public function getInstanceType($instanceName)
      */
     public function getPreference($type)
     {
+        $type = ltrim($type, '\\');
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/ObjectManager/Config/Compiled.php:123]: This does not look like a part of in-object cache, which is the purpose of this PR

File: lib/internal/Magento/Framework/ObjectManager/ObjectManager.php [hunk 4]
```diff
@@ -53,7 +53,6 @@ public function __construct(FactoryInterface $factory, ConfigInterface $config,
      */
     public function create($type, array $arguments = [])
     {
-        $type = ltrim($type, '\\');
         return $this->_factory->create($this->_config->getPreference($type), $arguments);
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/ObjectManager/ObjectManager.php:56]: This does not look like a part of in-object cache, which is the purpose of this PR

File: lib/internal/Magento/Framework/Code/Reader/ClassReader.php [hunk 5]
```diff
@@ -56,27 +61,32 @@ public function getConstructor($className)
...
-                $parentInterfaces = class_implements($parentClass);
-                if ($parentInterfaces) {
-                    $result = array_values(array_diff($interfaces, $parentInterfaces));
-                } else {
-                    $result = array_values($interfaces);
-                }
-            }
-            array_unshift($result, $parentClass);
+        if (isset($this->_classNameParents[$className])) {
+            return $this->_classNameParents[$className];
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/Code/Reader/ClassReader.php]: Please consider updating tests to verify this case.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
@sivajik34 Thank you for the PR. Idea looks good overall, however please check review comments.
Some numbers would be good as well.

File: lib/internal/Magento/Framework/ObjectManager/Config/Compiled.php (hunk 3 above)
Comment (adragus-inviqa) [reply lib/internal/Magento/Framework/ObjectManager/Config/Compiled.php:123]: I think OP has mixed up branches a bit: https://github.com/magento/magento2/pull/9189#issuecomment-292954874

File: lib/internal/Magento/Framework/Code/Reader/ClassReader.php (hunk 1 above)
Comment (sivajik34) [reply lib/internal/Magento/Framework/Code/Reader/ClassReader.php]: @ishakhsuvarov so can i use private $classNameParents?

File: lib/internal/Magento/Framework/ObjectManager/ObjectManager.php (hunk 4 above)
Comment (sivajik34) [reply lib/internal/Magento/Framework/ObjectManager/ObjectManager.php:56]: @ishakhsuvarov can i remove these changes and sent only changes related to this PR?

File: lib/internal/Magento/Framework/Code/Reader/ClassReader.php (hunk 1 above)
Comment (ishakhsuvarov) [reply lib/internal/Magento/Framework/Code/Reader/ClassReader.php]: Yes, that is what I meant :)

File: lib/internal/Magento/Framework/ObjectManager/ObjectManager.php (hunk 4 above)
Comment (ishakhsuvarov) [reply lib/internal/Magento/Framework/ObjectManager/ObjectManager.php:56]: Yes, that would be perfect.

**Comment by ishakhsuvarov** [comment]
Hi @sivajik34 
Are you going to proceed with this PR?
Thanks!

**Comment by sivajik34** [comment]
@ishakhsuvarov if its not that much helpful we can close it.no issues.

**Comment by ishakhsuvarov** [comment]
@sivajik34 
Ok. Thank you for collaboration

**Pull request closed without merging**